### PR TITLE
[dagster-dlift] project-level asset spec methods

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Project-level methods to retrieve asset specs.
## How I Tested These Changes
New unit tests.

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25485?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25486** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25486?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25485** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25485?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25484** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25484?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25483** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25483?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25482** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25482?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25481** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25481?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25458** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25457** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25446** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25446?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25444** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25443** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25443?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-dlift/dagster_dlift/project.py [hunk 1]
```diff
@@ -41,3 +41,27 @@ def compute_data(self) -> "DbtCloudProjectEnvironmentData":
...
+            *data.models_by_unique_id.values(),
+            *data.sources_by_unique_id.values(),
+            *data.tests_by_unique_id.values(),
+        ]
+        return [translator.get_spec(data) for data in all_external_data]
+
+    def get_asset_specs(
+        self, translator_cls: Type[DagsterDbtCloudTranslator] = DagsterDbtCloudTranslator
+    ) -> Sequence[AssetSpec]:
+        return [spec for spec in self.get_specs(translator_cls) if isinstance(spec, AssetSpec)]
```
Comment (benpankow) [inline examples/experimental/dagster-dlift/dagster_dlift/project.py:62]: should we cache `get_specs`, or have `get_specs` call these instead? not a huge deal, but does mean we run `get_spec(data)` for all data entries 2x if a user calls `get_asset_specs()` once and `get_check_specs()` once

File: examples/experimental/dagster-dlift/dagster_dlift/translator.py [hunk 2]
```diff
@@ -29,6 +30,15 @@ class DbtCloudContentData:
     content_type: DbtCloudContentType
     properties: Mapping[str, Any]
 
+    def from_raw_gql(self, raw_data: Mapping[str, Any]) -> "DbtCloudContentData":
+        content_type = DbtCloudContentType(raw_data["resourceType"].upper())
+        return DbtCloudContentData(content_type=content_type, properties=raw_data["properties"])
+
+
+def clean_asset_name(name: str) -> str:
```
Comment (benpankow) [inline examples/experimental/dagster-dlift/dagster_dlift/translator.py:38]: we really ought to stick this in the framework somewhere

File: examples/experimental/dagster-dlift/dagster_dlift/translator.py [hunk 3]
```diff
@@ -44,15 +54,35 @@ def environment_data(self) -> DbtCloudProjectEnvironmentData:
...
+            else:
+                # Macros, etc. which we don't create assets for.
+                continue
+        # Haven't actually figured out if it's possible for there to be more than one "asset-creatable" entity per test
+        parent_specs = cast(
+            Sequence[AssetSpec], [self.get_spec(data) for data in associated_data_per_parent]
+        )
+        return AssetCheckSpec(
+            name=clean_asset_name(data.properties["uniqueId"]),
+            asset=parent_specs[0].key,
```
Comment (benpankow) [inline examples/experimental/dagster-dlift/dagster_dlift/translator.py:83]: is this arbitrary, or is there meaning to the parent order in the gql response? seems fine, just might be worth noting if this is a framework feature on their end or just something we do to satisfy "check needs a single linked asset"

**Review by benpankow — APPROVED** [review]

File: examples/experimental/dagster-dlift/dagster_dlift/translator.py (hunk 3 above)
Comment (dpeng817) [reply examples/experimental/dagster-dlift/dagster_dlift/translator.py:83]: something we do to satisfy "check needs a single linked asset." See the above comment, but I'm not actually sure if it's possible for there to be more than one parent model/source for a given test. It seems like not, it's just a quirk of the data model.

File: examples/experimental/dagster-dlift/dagster_dlift/project.py (hunk 1 above)
Comment (dpeng817) [reply examples/experimental/dagster-dlift/dagster_dlift/project.py:62]: Yea I buy that we should cache it.

File: examples/experimental/dagster-dlift/dagster_dlift/translator.py (hunk 2 above)
Comment (dpeng817) [reply examples/experimental/dagster-dlift/dagster_dlift/translator.py:42]: :nod2:

Comment (dpeng817) [reply examples/experimental/dagster-dlift/dagster_dlift/translator.py]: I ended up adding it to `dagster._utils.names`

**Comment by dpeng817** [comment]
### Merge activity

* **Nov 1, 10:24 AM PDT**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/25485).

**Pull request merged**